sklearn/datasets/openml.py

```diff
@@ -63,7 +82,15 @@ def _open_openml_url(openml_path, data_home):
         try:
             with gzip.GzipFile(local_path, 'wb') as fdst:
                 fsrc = urlopen(_OPENML_PREFIX + openml_path)
-                shutil.copyfileobj(fsrc, fdst)
+                content = fsrc.read()
+
+                if md5_checksum is not None:
+                    md5_hash = hashlib.md5(content).hexdigest()
+                    if md5_checksum != md5_hash:
+                        warn('Data set file hash {} does not match the'
+                             ' checksum {}.'.format(md5_hash, md5_checksum))
+
+                fdst.write(content)
```

Comment:
This is not strictly equivalent. In `copyfileobj`,

> by default the data is read in chunks to avoid uncontrolled memory consumption

I'm not fully sure whether hashing by chunks (see e.g. these [SO answers](https://stackoverflow.com/questions/1131220/get-md5-hash-of-big-files-in-python)) is worthwhile here

Reply:
I'll do that, it might be necessary for very large data files.